### PR TITLE
[tune] remove dependency on gpy

### PR DESCRIPTION
**Pull request by matthewdeng**
## Description

Replace GPy with scikit-learn's `GaussianProcessRegressor` in the PB2 (Population Based Bandits) scheduler. 

 ## Additional information
  - Rewrote `TV_SquaredExp` kernel to implement sklearn's `Kernel` interface instead of GPy's `Kern`
  - Replaced `GPy.models.GPRegression` with `sklearn.gaussian_process.GaussianProcessRegressor`
  - Removed `gpy` from `tune-test-requirements.txt`
  - Updated documentation to only require scikit-learn (which is already a Ray dependency)

## Testing

Re-ran existing PB2 tests in `est_trial_scheduler_pbt.py` to validate change.

### CODE REVIEW

File: python/ray/tune/schedulers/pb2.py [hunk 1]
```diff
@@ -122,26 +120,20 @@ def _select_config(
...
     try:
-        m = GPy.models.GPRegression(X, y, kernel)
-    except np.linalg.LinAlgError:
-        # add diagonal ** we would ideally make this something more robust...
-        X += np.eye(X.shape[0]) * 1e-3
-        m = GPy.models.GPRegression(X, y, kernel)
-
-    try:
-        m.optimize()
+        m = GaussianProcessRegressor(
```
Comment (TimothySeah) [inline python/ray/tune/schedulers/pb2.py:126]: Is `m = GPy.models.GPRegression(X, y, kernel)` equivalent to 

```
m = GaussianProcessRegressor(
    kernel=kernel, optimizer="fmin_l_bfgs_b", alpha=1e-10
)
m.fit(X, y)
```

? Might be worth doublechecking if this matters.

File: python/ray/tune/schedulers/pb2_utils.py [hunk 2]
```diff
@@ -1,70 +1,104 @@
...
-class TV_SquaredExp(Kern):
+class TV_SquaredExp(Kernel):
     """Time varying squared exponential kernel.
     For more info see the TV-GP-UCB paper:
     http://proceedings.mlr.press/v51/bogunovic16.pdf
     """
 
     def __init__(
-        self, input_dim, variance=1.0, lengthscale=1.0, epsilon=0.0, active_dims=None
+        self,
```
Comment (TimothySeah) [inline python/ray/tune/schedulers/pb2_utils.py:16]: Do `input_dim` and `active_dims` no longer matter?

File: python/ray/tune/schedulers/pb2_utils.py [hunk 3]
```diff
@@ -1,70 +1,104 @@
...
+    @property
+    def hyperparameter_epsilon(self):
+        return Hyperparameter("epsilon", "numeric", self.epsilon_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        X = np.atleast_2d(X)
+        if Y is None:
+            Y = X
+
+        epsilon = np.clip(self.epsilon, 1e-5, 0.5)
```
Comment (TimothySeah) [inline python/ray/tune/schedulers/pb2_utils.py:48]: Where did the 1e-5 lower bound come from?

File: python/ray/tune/schedulers/pb2_utils.py [hunk 4]
```diff
@@ -1,70 +1,104 @@
...
+        return Hyperparameter("lengthscale", "numeric", self.lengthscale_bounds)
+
+    @property
+    def hyperparameter_epsilon(self):
+        return Hyperparameter("epsilon", "numeric", self.epsilon_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        X = np.atleast_2d(X)
+        if Y is None:
+            Y = X
```
Comment (TimothySeah) [inline python/ray/tune/schedulers/pb2_utils.py:46]: Do we need to do `Y = np.copy(X)` since it was `X2 = np.copy(X)` before?

File: python/ray/tune/schedulers/pb2_utils.py [hunk 5]
```diff
@@ -1,70 +1,104 @@
...
         )
 
-        return RBF * timekernel
+        K = rbf * timekernel
 
-    def Kdiag(self, X):
-        return self.variance * np.ones(X.shape[0])
+        if eval_gradient:
+            K_gradient_variance = K
+            dist2 = np.square(euclidean_distances(X_spatial, Y_spatial))
```
Comment (TimothySeah) [inline python/ray/tune/schedulers/pb2_utils.py:67]: Not sure why this is no longer divided by lengthscale - see the previous code: `dist2 = np.square(euclidean_distances(X, X2)) / self.lengthscale`

Iiuc __call__ in the scikitlearn implementation is similar to K + update_gradients_full in the GPy implementation, but the update_gradients_full portion seems different.

File: python/ray/tune/schedulers/pb2_utils.py [hunk 6]
```diff
@@ -171,13 +202,11 @@ def select_length(Xraw, yraw, bounds, num_f):
...
-                input_dim=X.shape[1], variance=1.0, lengthscale=1.0, epsilon=0.1
-            )
-            m = GPy.models.GPRegression(X, y, kernel)
-            m.optimize(messages=True)
+            kernel = TV_SquaredExp(variance=1.0, lengthscale=1.0, epsilon=0.1)
+            m = GaussianProcessRegressor(kernel=kernel, optimizer="fmin_l_bfgs_b")
+            m.fit(X, y)
 
-            scores.append(m.log_likelihood())
+            scores.append(m.log_marginal_likelihood_value_)
```
Comment (TimothySeah) [inline python/ray/tune/schedulers/pb2_utils.py:209]: Log likelihood isn't quite the same as log marginal likelihood but maybe that doesn't affect this algorithm in a meaningful way?

**Review by TimothySeah — COMMENTED** [review]
I'm not super familiar with gaussian processes or their implementations in scikitlearn/GPy. Also the scikitlearn -> GPy migration does not appear to be a 1 for 1 replacement. It might be worth getting a review from someone more familiar and/or having a good testing plan to ensure no regressions.

File: python/ray/tune/schedulers/pb2.py (hunk 1 above)
Comment (matthewdeng) [reply python/ray/tune/schedulers/pb2.py:126]: Yeah, these are the default values of `GaussianProcessRegressor`

File: python/ray/tune/schedulers/pb2_utils.py (hunk 2 above)
Comment (matthewdeng) [reply python/ray/tune/schedulers/pb2_utils.py:16]: Yeah sklearn `Kernel` doesn't need to pass it in during construction, it calculates at runtime.

File: python/ray/tune/schedulers/pb2_utils.py (hunk 4 above)
Comment (matthewdeng) [reply python/ray/tune/schedulers/pb2_utils.py:46]: No because we replaced in-place modification of

```python
        X = X[:, 1:]
        X2 = X2[:, 1:]
```

with

```python
        X_spatial = X[:, 1:]
        Y_spatial = Y[:, 1:]
```

File: python/ray/tune/schedulers/pb2_utils.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/tune/schedulers/pb2_utils.py:48]: It's consistent with `epsilon_bounds`, I think it should already be handled by sklearn but kept here for consistency with prev logic.

File: python/ray/tune/schedulers/pb2_utils.py (hunk 5 above)
Comment (matthewdeng) [reply python/ray/tune/schedulers/pb2_utils.py:67]: Yeah this is because sklearn uses logspace

File: python/ray/tune/schedulers/pb2_utils.py (hunk 6 above)
Comment (matthewdeng) [reply python/ray/tune/schedulers/pb2_utils.py:209]: gpy's implementation returns `_log_marginal_likelihood`, so it's the same thing.

**Review by TimothySeah — APPROVED** [review]
LGTM but might be worth adding testing notes to the PR description.
